src/utils/aws.py

```diff
@@ -0,0 +1,114 @@
+import logging
+import os
+import time
+from datetime import datetime
+
+import boto3
+
+
+class AWSClient:
+    def __init__(self, logger: logging.Logger):
+        """
+        Class to store AWS clients and methods for various AWS actions. Clients
+        are instantiated from AWS credentials passed via Compose secrets.
+
+        Attributes:
+            logger: Spark session logger that outputs to shared file in the
+                same format.
+            logs_client: Glue client connection for running crawlers.
+            glue_client: CloudWatch logs connection for uploading logs.
+            s3_client: S3 client connection for uploading extracted files.
+            s3_bucket: S3 bucket to upload extracts to.
+            s3_prefix: S3 path prefix within S3 bucket. Defaults to "iasworld".
+        """
+        self.logger = logger
+        self.logs_client = boto3.client("logs")
+        self.glue_client = boto3.client("glue")
+        self.s3_client = boto3.client("s3")
+        self.s3_bucket = os.getenv("AWS_S3_BUCKET")
+        self.s3_prefix = os.getenv("AWS_S3_PREFIX", "iasworld")
+
+    def run_and_wait_for_crawler(self, crawler_name) -> None:
+        initial_response = self.glue_client.get_crawler(Name=crawler_name)
+        if initial_response["Crawler"]["State"] == "READY":  # type: ignore
+            self.logger.info(f"Starting AWS Glue crawler {crawler_name}")
+            self.glue_client.start_crawler(Name=crawler_name)
+        else:
+            self.logger.warn(
+                f"AWS Glue crawler {crawler_name} is already running"
+            )
+            return
+
+        # Wait for the crawler to complete before triggering dbt tests
+        time_elapsed = 0
+        time_increment = 30
+        while True:
+            response = self.glue_client.get_crawler(Name=crawler_name)
+            state = response["Crawler"]["State"]  # type: ignore
+            if state in ["READY", "STOPPING"]:
+                self.logger.info(f"Crawler {crawler_name} has finished")
+                break
+            elif state == "RUNNING":
+                self.logger.info(
+                    (
+                        f"Crawler {crawler_name} is running: "
+                        f"{time_elapsed}s elapsed"
+                    )
+                )
+            time.sleep(time_increment)
+            time_elapsed += time_increment
```

Comment:
**[Suggestion, non-blocking]** I wonder if it's worth adding a generous timeout to this polling block, like 20 mins, just to avoid the possibility that the job hangs overnight if the crawler state transitions in an unexpected way? We could either raise an error if the job times out, or just print a warning:

```suggestion
        # Wait for the crawler to complete before triggering dbt tests
        time_elapsed = 0
        time_increment = 30
        timeout = 1200  # 20 minute timeout
        job_complete = False
        while time_elapsed < timeout:
            response = self.glue_client.get_crawler(Name=crawler_name)
            state = response["Crawler"]["State"]  # type: ignore
            if state in ["READY", "STOPPING"]:
                self.logger.info(f"Crawler {crawler_name} has finished")
                job_complete = True
                break
            elif state == "RUNNING":
                self.logger.info(
                    (
                        f"Crawler {crawler_name} is running: "
                        f"{time_elapsed}s elapsed"
                    )
                )
            time.sleep(time_increment)
            time_elapsed += time_increment
        if not job_complete:
            self.logger.warn(
                f"Crawler {crawler_name} was still running after the {timeout}s polling timeout; "
                "continuing execution without confirming its status"
            )
```